### PR TITLE
feat: allow to use an empty namespace

### PR DESCRIPTION
**Pull request by Jytesh**
@Kikobeats I am not sure how redis namespaces work, but I think empty namespaces will break the current redis store so can you update it?

### CODE REVIEW

**Comment by Kikobeats** [comment]
I'm not following the point of this PR, can you explain the change behind it? why not keep the original behavior?

**Comment by Jytesh** [comment]
Oh it just allows for empty namespace, original behaviour is unchanged, from https://github.com/lukechilds/keyv/pull/120

**Comment by Kikobeats** [comment]
title updated 🙂 looks like tests are failing

**Comment by Jytesh** [comment]
Yeah 
>>> I am not sure how redis namespaces work, but I think empty namespaces will break the current redis store so can you update it?

could you please update redis?

**Comment by Kikobeats** [comment]
Redis is a bit special related to namespaces, let me explain.

The thing is related to this line:

https://github.com/keyvhq/keyv/blob/empty%2Bnamespaces/packages/keyv-redis/src/index.js#L49

every time `.set` is called, an additional is called, the key is inserted into a Redis Set.

this is done because in case you want to clear all the keys, just you have to iterate the Set:

https://github.com/keyvhq/keyv/blob/empty%2Bnamespaces/packages/keyv-redis/src/index.js#L61

This is done in this way basically to avoid performing a Redis scan over all the keys because the Redis instance can have a lot of keys and a scan degrades the performance.

https://dzone.com/articles/the-effects-of-redis-scan-on-performance-and-how-k#:~:text=Redis%20Performance%20Degradation%20With%20SCAN,batches%20to%20minimize%20the%20effects.

so every time keys are prefixed using the `namespace`. What we can do there is, in case `namespace` is not provided, throw an error since in this case, I think it should be required.

**Comment by Jytesh** [comment]
so redis won't allow for empty namespace/no namespace?

**Comment by Kikobeats** [comment]
not really, you need a namespace prefix in order to clear the keys in a efficient way

**Comment by Jytesh** [comment]
We can prefix it as keyv, but then clearing would clear all keyv namespaces correct?

**Comment by Kikobeats** [comment]
Yes, that's the thing :(

Although I have to mention I never use `.clear`; I'm using `set` and `get` all the time and I don't think most people is going to clear the instance too often.

So maybe could be worth it to drop the current `clear` implementation and delegate the job to Redis UNLINK:

- https://rdbtools.com/blog/redis-delete-keys-matching-pattern-using-scan
- https://redis.io/commands/unlink

Doing that:

- It's less aggressive than a Redis scan
- It saves the extra request for `.set` and `.get`

What do you think?

**Comment by Jytesh** [comment]
Yes, sounds good

**Comment by Kikobeats** [comment]
give me one day, it needs a bit of rework 😇

**Comment by Jytesh** [comment]
Good luck! :D

File: packages/keyv/src/index.js [hunk 1]
```diff
@@ -25,7 +25,7 @@ class Keyv extends EventEmitter {
 			this.store.on('error', error => this.emit('error', error));
 		}
 
-		this.store.namespace = this.options.namespace;
+		this.store.namespace = this.options.namespace ? this.options.namespace + ':' : '';
```
Comment (Kikobeats) [inline packages/keyv/src/index.js:28]: IMHO `:` it's a separator and should be taken as internal detail, it not should be part of the namespace value.

In fact, you have the method `_getKeyPrefix` for getting namespace plus separator.

Comment (Jytesh) [reply packages/keyv/src/index.js:28]: aight

Comment (Jytesh) [reply packages/keyv/src/index.js:28]: You can change it, makes more sense to make it internal and not part of namespace

**Comment by Jytesh** [comment]
Can this be merged

**Pull request merged**